pnp-webpack-plugin cannot work outside of pnp runtime.

## Context
Trying to add yarn pnp support to cypress and ran into following error
```
**Details:** Error: Cannot find module 'pnpapi'
Require stack:
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/pnp-webpack-plugin/index.js
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/@cypress/webpack-batteries-included-preprocessor/index.js
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/lib/plugins/child/run_plugins.js
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/lib/plugins/child/index.js
Require stack:
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/pnp-webpack-plugin/index.js
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/@cypress/webpack-batteries-included-preprocessor/index.js
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/lib/plugins/child/run_plugins.js
- /Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/lib/plugins/child/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:972:15)
    at Function.external_module_.Module._resolveFilename (/Users/mluo/workspace/gitlab/ui-platform/projects/apps/px/.pnp.js:38306:54)
    at Module._load (internal/modules/cjs/loader.js:848:27)
    at external_module_.Module._load (/Users/mluo/workspace/gitlab/ui-platform/projects/apps/px/.pnp.js:38162:36)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12738)
    at Module.require (internal/modules/cjs/loader.js:1032:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at getSourceLocation (/Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/pnp-webpack-plugin/index.js:39:15)
    at makeResolver (/Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/pnp-webpack-plugin/index.js:53:26)
    at Object.<anonymous> (/Users/mluo/Library/Caches/Cypress/6.6.1/Cypress.app/Contents/Resources/app/packages/server/node_modules/pnp-webpack-plugin/index.js:147:10)
    at Module._compile (internal/modules/cjs/loader.js:1152:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1173:10)
    at Module.load (internal/modules/cjs/loader.js:992:32)
    at Module._load (internal/modules/cjs/loader.js:885:14)
    at external_module_.Module._load (/Users/mluo/workspace/gitlab/ui-platform/projects/apps/px/.pnp.js:38162:36)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12738)
```

Cypress server (an Electron app)is forking a new process to compile plugins from user project, however, the webpack config and compile code live with Cypress app. 

After chatting with merceyz in channel, it looks like pnp-webpack-plugin assumes it been located within the pnp project and it's not the case here.

https://discord.com/channels/226791405589233664/654372321225605128/822584013377241108

Would that be possible to port what we've done in webpack 5/enhanced-resolve so pnp-webpack-plugin can work in that scenario.

--- a/index.js
+++ b/index.js
@@ -5,16 +5,39 @@ function nothing() {
   // ¯\_(ツ)_/¯
 }
 
-function getModuleLocator(module) {
-  const pnp = require(`pnpapi`);
+/**
+ * get pnp api
+ * @param {*} source source path that may own by a pnp runtime
+ * @returns 
+ */
+function getPnpApi(source) {
+  let api;
+  try {
+    // try findPnpApi first
+    const m = require("module");
+    if (source && m.findPnpApi) {
+      api = m.findPnpApi(source);
+    }
+    // fallback to pnpapi
+    if (!api)
+      api = require(`pnpapi`);
+  } catch (ex) {
+    return null;
+  }
+  return api;
+}
 
+function getModuleLocator(module) {
   const moduleLocation = typeof module === `string`
     ? module
     : module.filename;
 
   if (!moduleLocation)
     throw new Error(`The specified module doesn't seem to exist on the filesystem`);
-
+  
+  const pnp = getPnpApi(moduleLocation);
+  if (!pnp)
+    return null;
   const moduleLocator = pnp.findPackageLocator(moduleLocation);
 
   if (!moduleLocator)
@@ -24,8 +47,9 @@ function getModuleLocator(module) {
 }
 
 function getDependencyLocator(sourceLocator, name) {
-  const pnp = require(`pnpapi`);
-
+  const pnp = getPnpApi();
+  if (!pnp) 
+    return null;
   const {packageDependencies} = pnp.getPackageInformation(sourceLocator);
   const reference = packageDependencies.get(name);
 
@@ -36,7 +60,9 @@ function getSourceLocation(sourceLocator) {
   if (!sourceLocator)
     return null;
 
-  const pnp = require(`pnpapi`);
+  const pnp = getPnpApi();
+  if (!pnp) 
+    return null;
 
   const sourceInformation = pnp.getPackageInformation(sourceLocator);
 
@@ -73,8 +99,6 @@ function makeResolver(sourceLocator, filter) {
 
     // Register a plugin that will resolve bare imports into the package location on the filesystem before leaving the rest of the resolution to Webpack
     resolver.getHook(`before-module`).tapAsync(`PnpResolver`, (requestContext, resolveContext, callback) => {
-      const pnp = require(`pnpapi`);
-
       let request = requestContext.request;
       let issuer = requestContext.context.issuer;
 
@@ -95,8 +119,8 @@ function makeResolver(sourceLocator, filter) {
 
       let resolutionIssuer = sourceLocation || issuer;
       let resolution;
-
       try {
+        const pnp = getPnpApi(resolutionIssuer);
         resolution = pnp.resolveToUnqualified(request, resolutionIssuer, {considerBuiltins: false});
       } catch (error) {
         if (resolveContext.missingDependencies)

--- a/index.js
+++ b/index.js
@@ -1,44 +1,20 @@
 const path = require(`path`);
 const {resolveModuleName} = require(`ts-pnp`);
+const {makeResolver} = require('./resolver');
 
 function nothing() {
   // ¯\_(ツ)_/¯
 }
 
-/**
- * get pnp api
- * @param {*} source source path that may own by a pnp runtime
- * @returns 
- */
-function getPnpApi(source) {
-  let api;
-  try {
-    // try findPnpApi first
-    const m = require("module");
-    if (source && m.findPnpApi) {
-      api = m.findPnpApi(source);
-    }
-    // fallback to pnpapi
-    if (!api)
-      api = require(`pnpapi`);
-  } catch (ex) {
-    return null;
-  }
-  return api;
-}
-
-function getModuleLocator(module) {
+function getModuleLocator(module, pnpapi) {
   const moduleLocation = typeof module === `string`
     ? module
     : module.filename;
 
   if (!moduleLocation)
     throw new Error(`The specified module doesn't seem to exist on the filesystem`);
-  
-  const pnp = getPnpApi(moduleLocation);
-  if (!pnp)
-    return null;
-  const moduleLocator = pnp.findPackageLocator(moduleLocation);
+
+  const moduleLocator = pnpapi.findPackageLocator(moduleLocation);
 
   if (!moduleLocator)
     throw new Error(`the specified module doesn't seem to be part of the dependency tree`);
@@ -46,136 +22,63 @@ function getModuleLocator(module) {
   return moduleLocator;
 }
 
-function getDependencyLocator(sourceLocator, name) {
-  const pnp = getPnpApi();
-  if (!pnp) 
-    return null;
-  const {packageDependencies} = pnp.getPackageInformation(sourceLocator);
+function getDependencyLocator(sourceLocator, name, pnpapi) {
+
+  const {packageDependencies} = pnpapi.getPackageInformation(sourceLocator);
   const reference = packageDependencies.get(name);
 
   return {name, reference};
 }
 
-function getSourceLocation(sourceLocator) {
-  if (!sourceLocator)
-    return null;
-
-  const pnp = getPnpApi();
-  if (!pnp) 
-    return null;
-
-  const sourceInformation = pnp.getPackageInformation(sourceLocator);
-
-  if (!sourceInformation)
-    throw new Error(`Couldn't find the package to use as resolution source`);
-
-  if (!sourceInformation.packageLocation)
-    throw new Error(`The package to use as resolution source seem to not have been installed - maybe it's a devDependency not installed in prod?`);
-
-  return sourceInformation.packageLocation.replace(/\/?$/, `/`);
-}
-
-function makeResolver(sourceLocator, filter) {
-  const sourceLocation = getSourceLocation(sourceLocator);
-
-  return resolver => {
-    const BACKWARD_PATH = /^\.\.([\\\/]|$)/;
-
-    const resolvedHook = resolver.ensureHook(`resolve`);
-
-    // Prevents the SymlinkPlugin from kicking in. We need the symlinks to be preserved because that's how we deal with peer dependencies ambiguities.
-    resolver.getHook(`file`).intercept({
-      register: tapInfo => {
-        return tapInfo.name !== `SymlinkPlugin` ? tapInfo : Object.assign({}, tapInfo, {fn: (request, resolveContext, callback) => {
-          callback();
-        }});
-      }
-    });
-
-    resolver.getHook(`after-module`).tapAsync(`PnpResolver`, (request, resolveContext, callback) => {
-      // rethrow pnp errors if we have any for this request
-      return callback(resolveContext.pnpErrors && resolveContext.pnpErrors.get(request.context.issuer));
-    });
-
-    // Register a plugin that will resolve bare imports into the package location on the filesystem before leaving the rest of the resolution to Webpack
-    resolver.getHook(`before-module`).tapAsync(`PnpResolver`, (requestContext, resolveContext, callback) => {
-      let request = requestContext.request;
-      let issuer = requestContext.context.issuer;
-
-      // When using require.context, issuer seems to be false (cf https://github.com/webpack/webpack-dev-server/blob/d0725c98fb752d8c0b1e8c9067e526e22b5f5134/client-src/default/index.js#L94)
-      if (!issuer) {
-        issuer = `${requestContext.path}/`;
-      // We only support issuer when they're absolute paths. I'm not sure the opposite can ever happen, but better check here.
-      } else if (!path.isAbsolute(issuer)) {
-        throw new Error(`Cannot successfully resolve this dependency - issuer not supported (${issuer})`);
-      }
-
-      if (filter) {
-        const relative = path.relative(filter, issuer);
-        if (path.isAbsolute(relative) || BACKWARD_PATH.test(relative)) {
-          return callback(null);
-        }
-      }
-
-      let resolutionIssuer = sourceLocation || issuer;
-      let resolution;
-      try {
-        const pnp = getPnpApi(resolutionIssuer);
-        resolution = pnp.resolveToUnqualified(request, resolutionIssuer, {considerBuiltins: false});
-      } catch (error) {
-        if (resolveContext.missingDependencies)
-          resolveContext.missingDependencies.add(requestContext.path);
-
-        if (resolveContext.log)
-          resolveContext.log(error.message);
-
-        resolveContext.pnpErrors = resolveContext.pnpErrors || new Map();
-        resolveContext.pnpErrors.set(issuer, error);
-
-        return callback();
-      }
-
-      resolver.doResolve(
-        resolvedHook,
-        Object.assign({}, requestContext, {
-          request: resolution,
-        }),
-        null,
-        resolveContext,
-        callback
-      );
-    });
-  };
-}
-
 module.exports = process.versions.pnp ? {
-  apply: makeResolver(null),
+  apply: makeResolver({pnpapi: require(`pnpapi`)}),
 } : {
   apply: nothing,
 };
 
 module.exports.makePlugin = (locator, filter) => process.versions.pnp ? {
-  apply: makeResolver(locator, filter),
+  apply: makeResolver({sourceLocator: locator, filter, pnpapi: require(`pnpapi`)}),
 } : {
   apply: nothing,
 };
 
-module.exports.moduleLoader = module => process.versions.pnp ? {
-  apply: makeResolver(getModuleLocator(module)),
-} : {
-  apply: nothing,
+module.exports.moduleLoader = (module) => {
+  if (process.versions.pnp) {
+    const pnpapi = require(`pnpapi`);
+    return {
+      apply: makeResolver({
+        sourceLocator: getModuleLocator(module, pnpapi),
+        pnpapi,
+      }),
+    };
+  }
+  return {
+    apply: nothing,
+  };
 };
 
 module.exports.topLevelLoader = process.versions.pnp ? {
-  apply: makeResolver({name: null, reference: null}),
+  apply: makeResolver({sourceLocator: {name: null, reference: null}, pnpapi: require(`pnpapi`)}),
 } : {
   apply: nothing,
 };
 
-module.exports.bind = (filter, module, dependency) => process.versions.pnp ? {
-  apply: makeResolver(dependency ? getDependencyLocator(getModuleLocator(module), dependency) : getModuleLocator(module), filter),
-} : {
-  apply: nothing,
+module.exports.bind = (filter, module, dependency) => {
+  if (process.versions.pnp) {
+    const pnpapi = require(`pnpapi`);
+    return {
+      apply: makeResolver({
+        sourceLocator: dependency
+          ? getDependencyLocator(getModuleLocator(module, pnpapi), dependency, pnpapi)
+          : getModuleLocator(module, pnpapi),
+        filter,
+        pnpapi,
+      }),
+    };
+  }
+  return {
+    apply: nothing,
+  };
 };
 
 module.exports.tsLoaderOptions = (options = {}) => process.versions.pnp ? Object.assign({}, options, {

--- a/resolver.js
+++ b/resolver.js
@@ -1,0 +1,117 @@
+const path = require(`path`);
+
+/**
+ * return source path given a locator
+ * @param {*} sourceLocator 
+ * @returns 
+ */
+function getSourceLocation(sourceLocator, pnpapi) {
+  if (!sourceLocator) return null;
+
+  const sourceInformation = pnpapi.getPackageInformation(sourceLocator);
+
+  if (!sourceInformation)
+    throw new Error(`Couldn't find the package to use as resolution source`);
+
+  if (!sourceInformation.packageLocation)
+    throw new Error(
+      `The package to use as resolution source seem to not have been installed - maybe it's a devDependency not installed in prod?`
+    );
+
+  return sourceInformation.packageLocation.replace(/\/?$/, `/`);
+}
+
+/**
+ *
+ * @param {*} sourceLocator
+ * @param {*} filter
+ * @returns
+ */
+function makeResolver(opts) {
+  const { sourceLocator, filter, pnpapi } = opts || {};
+  const sourceLocation = getSourceLocation(sourceLocator, pnpapi);
+  return (resolver) => {
+    const BACKWARD_PATH = /^\.\.([\\\/]|$)/;
+
+    const resolvedHook = resolver.ensureHook(`resolve`);
+
+    // Prevents the SymlinkPlugin from kicking in. We need the symlinks to be preserved because that's how we deal with peer dependencies ambiguities.
+    resolver.getHook(`file`).intercept({
+      register: (tapInfo) => {
+        return tapInfo.name !== `SymlinkPlugin`
+          ? tapInfo
+          : Object.assign({}, tapInfo, {
+              fn: (request, resolveContext, callback) => {
+                callback();
+              },
+            });
+      },
+    });
+
+    resolver
+      .getHook(`after-module`)
+      .tapAsync(`PnpResolver`, (request, resolveContext, callback) => {
+        // rethrow pnp errors if we have any for this request
+        return callback(
+          resolveContext.pnpErrors &&
+            resolveContext.pnpErrors.get(request.context.issuer)
+        );
+      });
+
+    // Register a plugin that will resolve bare imports into the package location on the filesystem before leaving the rest of the resolution to Webpack
+    resolver
+      .getHook(`before-module`)
+      .tapAsync(`PnpResolver`, (requestContext, resolveContext, callback) => {
+        let request = requestContext.request;
+        let issuer = requestContext.context.issuer;
+
+        // When using require.context, issuer seems to be false (cf https://github.com/webpack/webpack-dev-server/blob/d0725c98fb752d8c0b1e8c9067e526e22b5f5134/client-src/default/index.js#L94)
+        if (!issuer) {
+          issuer = `${requestContext.path}/`;
+          // We only support issuer when they're absolute paths. I'm not sure the opposite can ever happen, but better check here.
+        } else if (!path.isAbsolute(issuer)) {
+          throw new Error(
+            `Cannot successfully resolve this dependency - issuer not supported (${issuer})`
+          );
+        }
+
+        if (filter) {
+          const relative = path.relative(filter, issuer);
+          if (path.isAbsolute(relative) || BACKWARD_PATH.test(relative)) {
+            return callback(null);
+          }
+        }
+
+        let resolutionIssuer = sourceLocation || issuer;
+        let resolution;
+
+        try {
+          resolution = pnpapi.resolveToUnqualified(request, resolutionIssuer, {
+            considerBuiltins: false,
+          });
+        } catch (error) {
+          if (resolveContext.missingDependencies)
+            resolveContext.missingDependencies.add(requestContext.path);
+
+          if (resolveContext.log) resolveContext.log(error.message);
+
+          resolveContext.pnpErrors = resolveContext.pnpErrors || new Map();
+          resolveContext.pnpErrors.set(issuer, error);
+
+          return callback();
+        }
+
+        resolver.doResolve(
+          resolvedHook,
+          Object.assign({}, requestContext, {
+            request: resolution,
+          }),
+          null,
+          resolveContext,
+          callback
+        );
+      });
+  };
+}
+
+module.exports.makeResolver = makeResolver;

